docs/DeferredSignals.md

```diff
@@ -0,0 +1,271 @@
+  ; Load the page pointer from the CPUState
+  ldr x0, [x28, #(offsetof(CPUState, DeferredSignalHandlerPagePtr))]
+  ; Just store zero. (1 cycle plus no dependencies on a register. Super fast!)
+  ; Will store fine with no deferred signal, or SIGSEGV if there was one!
+  str xzr, [x0]
+```
+
+### Deferred signal handling
+In the case that FEX has received a signal, FEX's signal handler will first check to see if that thread's reference counter is zero or not.
+
+#### Reference counter is zero
+This is the easy case, just handle the signal as normal.
+
+#### Reference counter is not zero
+The signal handler now knows that FEX is in an uninterruptible code section. We check the signal to see if it is a synchronous signal or not.
+- If the signal is synchronous then we need to handle it as normal, because this is a hardware signal that we can't defer.
+- If it is an async signal (from tgkill, sigqueue, or something else) then we will start the deferring process.
+
+The deferring process starts with storing the kernel `siginfo_t` to a thread local array so we can restore it later.
+We then modify the permissions on the thread local `DeferredSignalHandlerPagePtr` to be `PROT_NONE`.
+We then immediately return from the signal handler so that FEX can resume its "uninterruptible" code section without breaking anything.
+Once the "uninterruptible" code section finishes, FEX will intentionally trigger a SIGSEGV by storing to the page.
+
+This will trigger a jump to FEX's SIGSEGV handler, where FEX will process the signal as if it was the previously deferred signal.
+the previous signal that was deferred.
+
+Once we are now handling the guest signal, FEX is in a guest signal frame and hence must be able to handle code compiling and other potentially
+awkward interactions without checking for additional signals. Handling further asynchronous signals in this vulnerable state requires special care.
+
+FEX-Emu is in a vulnerable state where any signals received will be deferred /and/ not handled at the end
+of "uninterruptible" code sections. This is because FEX is now currently in a guest signal frame and we need to handle code compiling and other
+potentially awkward interactions without checking for additional signals.
+
+Instead, before passing to the guest signal handler, FEX mprotects the thread local `DeferredSignalHandlerPagePtr` to be RW. The check for further
+signals is done by triggering a SIGILL after returning from the guest signal handler.
+
+- TODO: Overwrite mcontext_t things? I don't think this matters but there will be some private state that might leak SIGSEGV information?
+
+Once we are now handling the guest signal, FEX-Emu is in a vulnerable state where any signals received will be deferred /and/ not handled at the end
+of "uninterruptible" code sections. This is because FEX is now currently in a guest signal frame and we need to handle code compiling and other
+potentially awkward interactions without checking for additional signals.
```

Comment:
Information density is very thin in these paragraphs (almost looks like one paragraph was rewritten twice and the old versions weren't removed?). I've tried to condense it but I'm not getting the gist of what's actually happening here. Could you give this another pass?